habitat-lab/habitat/tasks/rearrange/actions/actions.py

```diff
@@ -386,6 +386,290 @@ def step(self, *args, is_last_action, **kwargs):
+        """
+        Set the goal of the agent
+        """
+        self.base_pose_goal = goal
+        if vel_goal is not None:
+            self.base_velocity_goal = vel_goal
+
+    def _compute_error_pose(self, base_pose, sim=None):
+        """
+        Updates error based on robot localization
+        """
+        base_pose_err = self._transform_global_to_base(self.base_pose_goal, base_pose, sim)
+
+        if not self.track_yaw:
+            base_pose_err[2] = 0.0
+
+        return base_pose_err
+
+    @staticmethod
+    def _velocity_feedback_control(base_pose_err, a, v_max):
+        """
+        Computes velocity based on distance from target.
+        Used for both linear and angular motion.
+
+        Current implementation: Trapezoidal velocity profile
+        """
+        t = np.sqrt(2.0 * abs(base_pose_err) / a)
+        v = min(a * t, v_max)
+        return v * np.sign(base_pose_err)
+
+    @staticmethod
+    def _turn_rate_limit(lin_err, heading_diff, w_max, tol=0.0):
+        """
+        Compute velocity limit that prevents path from overshooting goal
+
+        heading error decrease rate > linear error decrease rate
+        (w - v * np.sin(phi) / D) / phi > v * np.cos(phi) / D
+        v < (w / phi) / (np.sin(phi) / D / phi + np.cos(phi) / D)
+        v < w * D / (np.sin(phi) + phi * np.cos(phi))
+
+        (D = linear error, phi = angular error)
+        """
+        assert lin_err >= 0.0
+        assert heading_diff >= 0.0
+
+        max_heading_ang = np.pi / 10
+        if heading_diff > max_heading_ang:
+            return 0.0
+        else:
+            return (
+                w_max
+                * lin_err
+                / (
+                    np.sin(heading_diff)
+                    + heading_diff * np.cos(heading_diff)
+                    + 1e-5
+                )
+            )
+
+    def _feedback_simple(self, base_pose_err):
+        """
+        Feedback controllers
+        """
+        v_cmd = w_cmd = 0
+
+        lin_err_abs = np.linalg.norm(base_pose_err[0:2])
+        ang_err = base_pose_err[2]
+
+        # Go to goal XY position if not there yet
+        if lin_err_abs > self.lin_error_tol:
+            heading_err = np.arctan2(base_pose_err[1], base_pose_err[0])
+            heading_err_abs = abs(heading_err)
+
+            if abs(heading_err_abs - np.pi) <= 0.01:
+                heading_err_abs = 0
+                heading_err = 0
+
+            # Compute linear velocity and allow the agent to move backward
+            v_raw = self._velocity_feedback_control(
+                np.sign(base_pose_err[0])*lin_err_abs, self.acc_lin, self.v_max
+            )
+            v_limit = self._turn_rate_limit(
+                lin_err_abs,
+                heading_err_abs,
+                self.w_max / 2.0,
+                tol=self.lin_error_tol,
+            )
+            # Allow the agent to move backward
+            v_cmd = np.clip(v_raw, -v_limit, v_limit)
```

Comment:
why does this allow agent to move backward?

Reply:
or i guess why is that the comment ,seems to clip it to either backward or forward

Reply:
I fixed the issue here. This is not the part that allows the agent to move backward. It should be the one that allows the non-absolute value to pass